### PR TITLE
Stack overflow when searching a bk tree

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/10813

The issue here is that BKtrees can perform badly when provided with some pathological assemblies.  Specifically, in the case of obfuscated dlls we might get 10k symbols that are all have a single unicode character as their name.  The BK tree will then view these all as unique, but all with an edit distance of one from each other.  This will form a chain that is 10k long which will then blow the stack when we attempt to search for an item that is only an edit distnace of 1 away from these guys (i.e. any single character).

The "fix" is twofold.  First, for metadata assemblies we don't even load internal symbols anymore.  It was highly unlikely that the user could use these symbols anyways, so loading them wasn't that helpful to begin with. This will address the obfuscated code issue.  However, there's still an issue where technically someone could make such a pathological situation in normal source code.  To prevent that code from being an issue, we also cap the recursion amount when searching the BK tree.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Fixes https://github.com/dotnet/roslyn/issues/10813

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs [hunk 1]
```diff
@@ -278,7 +278,12 @@ private int BinarySearch(string name)
                     return info;
                 }
 
-                info = await LoadOrCreateSymbolTreeInfoAsync(solution, assembly, reference.FilePath, loadOnly, cancellationToken).ConfigureAwait(false);
+                // We don't include internals from metadata assemblies.  It's less likely that
+                // a project would have IVT to it and so it helps us save on memory.  It also
+                // means we can avoid loading lots and lots of obfuscated code in the case hte
```
Comment (brettfo) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs]: `hte` -> `the`

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs [hunk 2]
```diff
@@ -387,44 +396,65 @@ private static int CompareNodes(Node x, Node y, IReadOnlyList<Node> nodeList)
...
 
             foreach (var grouping in membersByName)
             {
-                GenerateNodes(grouping.Key, nodeIndex, grouping, list);
+                GenerateNodes(grouping.Key, nodeIndex, grouping, list, lookup);
             }
         }
 
-        private static Func<ISymbol, IEnumerable<ISymbol>> s_getMembers = symbol =>
+        private static Func<ISymbol, IEnumerable<ISymbol>> s_getMembersNoPrivate = symbol =>
```
Comment (brettfo) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs]: `s_getMembersNotPrivate` ('No' -> 'Not')

File: src/Workspaces/Core/Portable/Utilities/BKTree.cs [hunk 3]
```diff
@@ -97,8 +97,28 @@ public IList<string> Find(string value, int? threshold = null)
...
+            Node currentNode,
+            char[] queryCharacters,
+            int queryLength,
+            int threshold,
+            List<string> result, 
+            int recursionCount)
         {
+            // Don't bother recursing too deeply in the case of pathological trees.
+            // This really only happens when the actual code is strange (like
+            // 10,000 symbols all a single letter long).  In htat case, searching
```
Comment (brettfo) [inline src/Workspaces/Core/Portable/Utilities/BKTree.cs]: `that`

File: src/Workspaces/Core/Portable/Utilities/BKTree.cs [hunk 4]
```diff
@@ -97,8 +97,28 @@ public IList<string> Find(string value, int? threshold = null)
...
+            // down this path will be fairly fruitless anyways.
+            //
+            // Note: this won't affect good searches against good data even if this
+            // pathological chain exists.  That's because the good items will still
+            // cluster near the root node in the tree, and won't be off the end of
+            // this long chain.
+            if (recursionCount > 256)
```
Comment (brettfo) [inline src/Workspaces/Core/Portable/Utilities/BKTree.cs]: Constant for easier tuning.  Also, is there a more deterministic way to decide if we're about to SO?

File: src/Workspaces/Core/Portable/Utilities/SpellChecker.cs [hunk 5]
```diff
@@ -11,7 +11,7 @@ namespace Roslyn.Utilities
 {
     internal class SpellChecker
     {
-        private const string SerializationFormat = "1";
+        private const string SerializationFormat = "2";
```
Comment (brettfo) [inline src/Workspaces/Core/Portable/Utilities/SpellChecker.cs]: Is this just to invalidate old trees with 10k single-character items?  If so, won't the recursion check above prevent the SO?

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Utilities/SpellChecker.cs]: Yes.  This will help us free up memory as well which is nice.

**Comment by Pilchie** [comment]
:+1:

**Comment by brettfo** [comment]
Pending comment typos I'm satisfied: :shipit:

**Comment by dpoeschl** [comment]
:+1:

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo.cs]: Actually, 'No' was intentional here.

**Comment by CyrusNajmabadi** [comment]
Fixes #10813

**Pull request merged**